### PR TITLE
planner: ignore index_merge when partial path has global index

### LINKED ISSUE #43013: The global index not work with IndexMerge executor
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
set tidb_enable_global_index = true;
mysql> show create table tpk2;
+-------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                                                                                                                                                                                                                             |
+-------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tpk2  | CREATE TABLE `tpk2` (
  `a` int(11) DEFAULT NULL,
  `b` int(11) DEFAULT NULL,
  `c` int(11) NOT NULL,
  `d` int(11) NOT NULL AUTO_INCREMENT,
  KEY `idx_bc` (`b`,`c`),
  UNIQUE KEY `uidx_ac` (`a`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin AUTO_INCREMENT=30002
PARTITION BY HASH (`c`) PARTITIONS 2 |
+-------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

insert into tpk2 values (1, 1, 1, 1), (2, 2, 2, 2);

mysql> select /*+ use_index_merge(tpk2, uidx_ac, idx_bc) */ * from tpk2 where a=1 or b=2;
+------+------+---+---+
| a    | b    | c | d |
+------+------+---+---+
|    2 |    2 | 2 | 2 |
+------+------+---+---+
1 row in set (0.01 sec)

mysql> select * from tpk2 where a=1 or b=2;
+------+------+---+---+
| a    | b    | c | d |
+------+------+---+---+
|    2 |    2 | 2 | 2 |
|    1 |    1 | 1 | 1 |
+------+------+---+---+
2 rows in set (0.01 sec)
```

### 2. What did you expect to see? (Required)

The result is same with using indexMerge or not.

### 3. What did you see instead (Required)

The results are different.

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

### PR DESCRIPTION
**Pull request by L-maple**
### What problem does this PR solve?

Issue Number: ref #43013 

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: executor/partition_table_test.go [hunk 1]
```diff
@@ -4342,6 +4343,36 @@ func TestGlobalIndexForIssue40149(t *testing.T) {
...
+		KEY idx_bd (b, c),
+		UNIQUE KEY uidx_ac(a)
+	) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin PARTITION BY RANGE (c) (
+ 	PARTITION p0 VALUES LESS THAN (10),
+ 	PARTITION p1 VALUES LESS THAN (MAXVALUE));`)
+	tk.MustExec("insert into t values (1,1,1,1),(2,2,2,2),(3,3,3,3),(4,4,4,4),(5,5,5,5),(6,6,6,6),(7,7,7,7),(8,8,8,8);")
+	tk.MustExec("analyze table t")
+	// when index_merge has global index as its partial path, ignore it.
+	assert.False(t, tk.MustUseIndex("select /*+ use_index_merge(t, uidx_ac, idx_bc) */ * from t where a=1 or b=2", "uidx_ac"))
+	tk.MustQuery("select /*+ use_index_merge(t, uidx_ac, idx_bc) */ * from t where a=1 or b=2").Check(
```
Comment (Defined2014) [inline executor/partition_table_test.go:4372]: Should sort the result to keep CI stable.

File: planner/core/indexmerge_path.go [hunk 2]
```diff
@@ -339,6 +339,14 @@ func (ds *DataSource) buildIndexMergeOrPath(
 	indexMergePath := &util.AccessPath{PartialIndexPaths: partialPaths}
 	indexMergePath.TableFilters = append(indexMergePath.TableFilters, filters[:current]...)
 	indexMergePath.TableFilters = append(indexMergePath.TableFilters, filters[current+1:]...)
+	// If global index exists, index merge is not allowed.
+	// Global index is not compatible with IndexMergeReaderExecutor.
+	for i := range partialPaths {
+		if partialPaths[i].Index != nil && partialPaths[i].Index.Global {
```
Comment (Defined2014) [inline planner/core/indexmerge_path.go:345]: return a warning msg is better here

**Review by Defined2014 — COMMENTED** [review]
Please not close issue, you can use `ref` instead of `close`

**Comment by L-maple** [comment]
> Please not close issue, you can use `ref` instead of `close`

OK, I change the issue mark `close` to `ref`.

File: executor/partition_table_test.go [hunk 3]
```diff
@@ -4342,6 +4343,36 @@ func TestGlobalIndexForIssue40149(t *testing.T) {
...
+		d int(11) NOT NULL AUTO_INCREMENT,
+		KEY idx_bd (b, c),
+		UNIQUE KEY uidx_ac(a)
+	) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin PARTITION BY RANGE (c) (
+ 	PARTITION p0 VALUES LESS THAN (10),
+ 	PARTITION p1 VALUES LESS THAN (MAXVALUE));`)
+	tk.MustExec("insert into t values (1,1,1,1),(2,2,2,2),(3,3,3,3),(4,4,4,4),(5,5,5,5),(6,6,6,6),(7,7,7,7),(8,8,8,8);")
+	tk.MustExec("analyze table t")
+	// when index_merge has global index as its partial path, ignore it.
+	assert.False(t, tk.MustUseIndex("select /*+ use_index_merge(t, uidx_ac, idx_bc) */ * from t where a=1 or b=2", "uidx_ac"))
```
Comment (Defined2014) [inline executor/partition_table_test.go:4371]: use `require.Flase` here

Comment (L-maple) [reply executor/partition_table_test.go:4371]: resolved

**Review by mjonss — APPROVED** [review]

**Review by Defined2014 — APPROVED** [review]

**Pull request merged**